lib/active_record/virtual_attributes/virtual_total.rb

```diff
@@ -109,26 +109,26 @@ def virtual_aggregate_arel(reflection, method_name, column)
                           reflection.klass.arel_attribute(column).send(method_name)
                         end
 
-          # query: SELECT * FROM foreign_table
-          query = if reflection.scope
-                    reflection.klass.instance_exec(nil, &reflection.scope)
-                  else
-                    reflection.klass.all
-                  end
-
-          # query: SELECT [COUNT(*)] FROM foreign_table
-          query = query.select(arel_column)
-
-          # query: SELECT COUNT(*) FROM foreign_table [WHERE main_table.id = foreign_table.id]
-          foreign_table = reflection.klass.arel_table
-          if ActiveRecord.version.to_s >= "5.1"
-            join_keys = reflection.join_keys
-          else
-            join_keys = reflection.join_keys(reflection.klass)
-          end
-          query       = query.except(:order).where(t[join_keys.foreign_key].eq(foreign_table[join_keys.key]))
-
-          sql         = query.to_sql
+          # query: SELECT COUNT(*) FROM main_table JOIN foreign_table ON main_table.id = foreign_table.id JOIN ...
+          relation_query   = joins(reflection.name).select(arel_column)
+          query            = relation_query.arel
+          bound_attributes = relation_query.bound_attributes
+
+          # algorithm:
+          # - remove main_table from this sub query. (it is already in the primary query)
+          # - move the foreign_table from the JOIN to the FROM clause
+          # - move the main_table.id = foreign_table.id from the ON clause to the WHERE clause
+
+          # query: SELECT COUNT(*) FROM main_table [ ] JOIN ...
+          join = query.source.right.shift
+          # query: SELECT COUNT(*) FROM [foreign_table] JOIN ...
+          query.source.left = join.left
+          # query: SELECT COUNT(*) FROM foreign_table JOIN ... [WHERE main_table.id = foreign_table.id]
+          query.where(join.right.expr)
+
+          # convert bind variables from ? to actual values. otherwise, sql is incomplete
+          conn = connection
+          sql  = conn.unprepared_statement { conn.to_sql(query, bound_attributes) }
```

Comment:
naive question... are `unprepared_statement` and `bound_attributes` public methods and work the same with rails 5.1, 5.2 and 6.0?

Reply:
This is taken from ActiveRecord `Relation#to_sql` as the official way to bind variables back into sql statements.

The `to_sql` statement (if you ignore moving code around) has been the same since before 4.0 and remains the same through 6.0.

thanks to @NickLaMuro for pointing me in the right direction here